Unified Testing Script

rutkovskii opened the pull request:
Completed:
1. Precipitation Metrics based on torchmetrics and based on the code from `calc_metrics_test_set.py`
2. UNetBase with test_step that uses Precipitation Metrics
3. `calc_metrics_test_set.py` – cleaned and using test_step, supports command line arguments to select test file, and also 3 different versions of output. 

I left a couple TODOs: 
1. To remove old code once we finalize this pull request
2. To integrate plotting (if needed)
3. (Maybe) you have idea on where to put `--threshold` argument or how to rework it in `UNetBase`. 

`test_step()` uses the method from the original `calc_metrics_test_set.py`
There is a minor difference on how it does it when compared to `test_precip_lightning.py`.

The result of updated `calc_metrics_test_set.py`: [model_metrics_0.5.txt](https://github.com/user-attachments/files/19033961/model_metrics_0.5.txt).

@HansBambel, please run the script with your trained models (make sure to change the names to match Camel Case) to confirm whether the metrics are performing correctly. I cannot access the ground truth (results from the old scripts).

Will resolve #9

rutkovskii commented:
@HansBambel No rush on checking it, just leaving a reminder in case you missed the notification about Review Request.

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,152 @@
+    test_dl = torch.utils.data.DataLoader(dataset, batch_size=1, num_workers=1, shuffle=False, pin_memory=True, persistent_workers=True)
+    
+    # Create trainer with device specification
+    trainer = pl.Trainer(logger=False, enable_checkpointing=False, accelerator='gpu' if torch.cuda.is_available() else 'cpu')
```
```suggestion
    trainer = pl.Trainer(logger=False, enable_checkpointing=False, accelerator=device)
```

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,152 @@
+    print(f"Using device: {device}")
+
+    # Set PyTorch to use higher precision for matrix multiplications
+    torch.set_float32_matmul_precision('high')
```
Why set this?

rutkovskii replied:
https://stackoverflow.com/q/76797190

This stack explains the warning well. It is not necessary. I just place it sometimes to remove the warning. I should remove it.

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,152 @@
+        else:
+            loaded_model = model.load_from_checkpoint(os.path.join(model_folder, model_file))
+        
+        loaded_model.to(device)
+        loaded_model.precip_metrics = PrecipitationMetrics(threshold=threshold).to(device)
```
Is the explicit move to the device needed when using pytorch lightning?

rutkovskii replied:
I  encountered issue related to the information nit being on the correct device for either persistence or for test_step. I will run check if it is still the case, and get back to you.

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,152 @@
+        print(f"Loading model: {model_name}")
 
+        if model_file == "PersistenceModel":
+            loaded_model = model(Namespace())
```
What is `Namespace()` doing?

rutkovskii replied:
For example: in training script, The [ArgumentParser.parse_args()](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args) method runs the parser and places the extracted data in a [argparse.Namespace](https://docs.python.org/3/library/argparse.html#argparse.Namespace) object, which then passed into the models to train. 

Since `PersistenceModel` is a derivation of `UNetBase` that asks for arguments, I had to provide an empty `Namespace` object that allows `PersistenceModel` to resort to using the defaults defined in `UNetBase`

HansBambel commented on metric/precipitation_metrics.py:
```diff
@@ -0,0 +1,135 @@
+            preds: Model predictions (normalized)
+            target: Ground truth (normalized)
+        """
+        # Move tensors to the same device as the metric states
+        preds = preds.to(self.device)
+        target = target.to(self.device)
```
I don't see that moving the tensors to the device explicitly is needed [in the example](https://lightning.ai/docs/torchmetrics/stable/pages/implement.html).

Is it actually needed?

rutkovskii replied:
It was related to the issue with the above:

https://github.com/HansBambel/SmaAt-UNet/pull/42#discussion_r1978308735

HansBambel commented on metric/precipitation_metrics.py:
```diff
@@ -0,0 +1,135 @@
+            preds_hourly = preds_denorm * 12
+            target_hourly = target_denorm * 12
+            
+            # Apply threshold to get binary masks
+            preds_mask = preds_hourly > self.threshold
+            target_mask = target_hourly > self.threshold
+            
+            # Compute confusion matrix
+            confusion = target_mask.view(-1) * 2 + preds_mask.view(-1)
+            bincount = torch.bincount(confusion, minlength=4)
+            
+            # Update confusion matrix states
+            self.total_tn += bincount[0]
+            self.total_fp += bincount[1]
+            self.total_fn += bincount[2]
+            self.total_tp += bincount[3]
```
These should also get updated when we not denormalize, right? Then, this should be unindented.

HansBambel reviewed (commented):
Sorry for only getting back to this now, but I was visiting family over the weekend.

Something that I noticed is that I needed to install tensorboard to be able to train, so we should add it to the regular dependencies.

Looking good for now, but I was not able to test it yet with my trained models as I am currently re-training. Afterwards, I will run the tests as well.

Something that I wondered: Should we also log the precipitation metrics in the training? Or would this hurt performance too much? I haven't tested it, but it would be nice to have this in addition to the regression metrics and losses.

rutkovskii commented:
ToDo:
1. Tensorboard to main deps
2. Precipitation metrics in training
3. Remove .to_device() and test if it works without it. If so clear out any arguments and parts.
4. Dedent when running not normalized

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,147 @@
+    """
+    results = {}
+
+    # Check for CUDA availability and set device
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    print(f"Using device: {device}")
```
This is not needed. I removed it and it was still loading with cuda. It even threw an error, because the trainer was expecting a string and not `torch.device()`

rutkovskii replied:
I sa it as well. Waiting for results of 2 testing processes and will make a push.

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,147 @@
+    test_dl = torch.utils.data.DataLoader(dataset, batch_size=1, num_workers=1, shuffle=False, pin_memory=True, persistent_workers=True)
+    
+    # Create trainer with device specification
+    trainer = pl.Trainer(logger=False, enable_checkpointing=False, accelerator=device)
```
```suggestion
    trainer = pl.Trainer(logger=False, enable_checkpointing=False)
```

HansBambel commented on calc_metrics_test_set.py:
```diff
@@ -1,143 +1,147 @@
+        else:
+            loaded_model = model.load_from_checkpoint(os.path.join(model_folder, model_file))
+        
+        loaded_model.to(device)
+        loaded_model.precip_metrics = PrecipitationMetrics(threshold=threshold).to(device)
```
I didn't need it
```suggestion
       loaded_model.precip_metrics = PrecipitationMetrics(threshold=threshold)
```

rutkovskii commented:
**Completed:**

1. Removed manual handling of device
2. Plotting of all metrics (with handling of Nans from normalized mode"
3. Printing training metrics at the end of epoch (`self.log()`) and aggregated metrics (`print`) @HansBambel Please check the [log-J29512067.txt](https://github.com/user-attachments/files/19112425/log-J29512067.txt) and tell me if you want me to remove anything such as `Epoch 1 - TRAIN METRICS: mse: 0.0908 | mse_denorm: 207.7536 | mse_pixel: 0.0025 | precision: 0.4808 | recall: 0.6086 | accuracy: 0.5912 | f1: 0.5372 | csi: 0.3673 | far: 0.5192 | hss: 0.0900
`. See Point 4 from potential future improvements.
4. I updated the dependencies and `uv.lock`. @HansBambel Could you please try to run `uv sync --frozen` and confirm if it works on your system by seeing that the CUDA version of torch was downloaded? With my HPC and SLURM, ensuring whether it has been done properly and replicably on other personally owned machines is complicated.
5. Separated training, validation, test metrics – and had to overwrite standard test_metrics in the`calc_metrics_test_set.py` in order to pass `denormalize` flag because by default it is initialized as denormalize=True. 
6. Checked: training (on 1 and 2 gpus), `calc_metrics_test_set.py` with denormalized and normalized data as well with plotting and loading from a file (from txt and csv)



Example of logs for testing:
[log-J29510758.txt](https://github.com/user-attachments/files/19112408/log-J29510758.txt) – testing only "normalized" data without doing mse-denorm and mse-pixel
[log-J29510954.txt](https://github.com/user-attachments/files/19112417/log-J29510954.txt) – loading from a file



---
**Potential Future Improvements:**
1. Some supporting functions can be separated into separate file (parsing argument, converting tensors, saving metrics) and used from `utils/`
2. Adding additional flags for the parser for `input-file-type` and `output-file-type` so that person could load metrics of one type and just convert them to the other type. 
3. Or we can just save metrics in all formats at the same time (txt, json, csv)
4. If you want to keep printing of Aggregated metrics since it is repeated 3 times, I can put it into function.
```
        metrics_str = " | ".join([f"{name}: {value.item() if isinstance(value, Tensor) else value:.4f}" 
                                 for name, value in test_metrics_dict.items() 
                                 if not (isinstance(value, Tensor) and torch.isnan(value)) and 
                                    not (not isinstance(value, Tensor) and np.isnan(value))])
        print(f"\n\nTEST METRICS: {metrics_str}")
```

HansBambel commented:
3. Oh wow, that is noisy. Maybe it is better to just have these metrics at test time. What do you think?
4. Yep, works, but now matplotlib is also a required dependency for the test-script, so we should move this from `scripts`-group to the regular dependencies.

I think having the metrics in a json is the easiest for most purposes. I think having a converter function would be unnecessary. 

> If you want to keep printing of Aggregated metrics since it is repeated 3 times, I can put it into function.

Yes, that would be best :)

Thank you for all this!

rutkovskii commented:
3. I do agree it becomes overwhelming – slightly more readable when it happens in the terminal, but otherwise it is just makes the experience worse. So I will revert back to just printing loss without all metrics during training. But see Point 2 from next section.
4. I will move matplotlib to main group.

---
1. Do you want me to remove csv and txt file output? 
2. I still can keep `self.log` of all metrics at the end of each epoch and also print Aggregated metrics like so – but let me know if this is what you want:
```
metrics_str = " | ".join([f"{name}: {value.item() if isinstance(value, Tensor) else value:.4f}" 
                                 for name, value in test_metrics_dict.items() 
                                 if not (isinstance(value, Tensor) and torch.isnan(value)) and 
                                    not (not isinstance(value, Tensor) and np.isnan(value))])
        print(f"\n\nTEST METRICS: {metrics_str}")
````

HansBambel commented:
1. You can keep it
2. Yep, sounds good

rutkovskii commented:
This should be it.

HansBambel reviewed (approved):
Looks good! Thanks again!
